### ISSUE TITLE
Cypress running in parallel with TDE : Tasks hashes haven't been recorded.

### ISSUE THREAD

**Issue by roldengarm** [nrwl/nx#10302]
## Current Behavior
Running e2e tests in parallel using Cypress dashboard. Enabled NX Cloud in our CI pipeline. Getting this error:

`Error: Command execution failed (distributed task execution: ...). Tasks hashes haven't been recorded.`

Command:

`nx affected --base=origin/master --head=HEAD --target=e2e --maxParallel=2 --ci-build-id $(Build.BuildNumber) --record --key 'CYPRESS_KEY' -- --parallel`

## Expected Behavior
When running Cypress tests in parallel, it's expected that TDE still works. 

## Steps to Reproduce
See above.

### Failure Logs
See above.

### Environment
See also this bug: https://github.com/nrwl/nx/issues/8014

```
$ nx report

 >  NX   Report complete - copy this into the issue template

   Node : 16.13.2
   OS   : win32 x64
   npm  : 8.1.2

   nx : 13.7.1
   @nrwl/angular : 13.7.1
   @nrwl/cli : 13.7.1
   @nrwl/cypress : 13.7.1
   @nrwl/detox : undefined
   @nrwl/devkit : 13.7.1
   @nrwl/eslint-plugin-nx : 13.7.1
   @nrwl/express : undefined
   @nrwl/jest : 13.7.1
   @nrwl/js : undefined
   @nrwl/linter : 13.7.1
   @nrwl/nest : undefined
   @nrwl/next : undefined
   @nrwl/node : undefined
   @nrwl/nx-cloud : 14.0.3
   @nrwl/react : undefined
   @nrwl/react-native : undefined
   @nrwl/schematics : undefined
   @nrwl/storybook : 13.7.1
   @nrwl/tao : 13.7.1
   @nrwl/web : undefined
   @nrwl/workspace : 13.7.1
   typescript : 4.5.5
   rxjs : 7.4.0
   ---------------------------------------
   Community plugins:
         @angular/animations: 13.2.0
         @angular/cdk: 13.3.1
         @angular/common: 13.2.0
         @angular/compiler: 13.2.0
         @angular/core: 13.2.0
         @angular/forms: 13.2.0
         @angular/material: 13.3.1
         @angular/material-date-fns-adapter: 13.3.1
         @angular/platform-browser: 13.2.0
         @angular/platform-browser-dynamic: 13.2.0
         @angular/router: 13.2.0
         @auth0/auth0-angular: 1.8.2
         @angular-devkit/build-angular: 13.2.1
         @angular/cli: 13.2.1
         @angular/compiler-cli: 13.2.0
         @angular/language-service: 13.2.0
         @currents/nx: 0.0.5
         @nx-dotnet/core: 1.8.0
         @testing-library/angular: 11.0.4
```

**Comment by StalkAltan** [comment]
Hi @roldengarm! Thanks for reporting this.

This is what will be output on the main job if one of your agents fails. Would it be possible for you to provide any logs from your individual agents as well?

**Comment by roldengarm** [comment]
@StalkAltan 

First, I've switched from Cypress to Currents and am using [@currents/nx](https://github.com/currents-dev/currents-nx) 
Same issue though.

Targets in my project.json are:

```
 "currents-test": {
      "executor": "@currents/nx:currents",
      "options": {
        "cypressExecutor": "test"
      }
    },
    "test": {
      "executor": "@nrwl/cypress:cypress",
      "options": {
        "cypressConfig": "apps/web-automation-tests/cypress.ui.component.json",
        "devServerTarget": "web:serve:development"
      },
      "configurations": {
        "production": {
          "devServerTarget": "web:serve:production"
        }
      }
    },
```
I'm running this command in the main job:

```
npx nx affected --base=origin/main --target=currents-test --group nx --record --key KEY --ci-build-id $(Build.BuildId) -- --parallel
```

The output from the main job is:
```
  Agent 'Agent 72642' exited with the error: Critical Error in Agent: "Command execution failed (distributed task execution: h9sMBNqnhlE). Tasks hashes haven't been recorded.".
```
The output of the agent job is:
```
>  NX   Running target currents-test for 1 project(s):

    - web-automation-tests
   
   With flags:
     --group=nx
     --record=true
     --key=...
     --ciBuildId=...

 >  NX   Successfully ran target currents-test for 1 projects

Error: Command execution failed (distributed task execution: h9sMBNqnhlE). Tasks hashes haven't been recorded.
    at readCompletedTasks (/home/vsts/work/1/s/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:6412)
    at invokeTasksUsingRunMany (/home/vsts/work/1/s/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:7468)
    at /home/vsts/work/1/s/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:5671
    at Generator.next (<anonymous>)
    at _0x2672e6 (/home/vsts/work/1/s/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:2777)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```
Please let me know if you  need anything else, I'd love to be able to use distributed caching for this.

**Comment by roldengarm** [comment]
This is still an issue @StalkAltan

**Comment by gund** [comment]
I can confirm this is an issue and it's not related to E2E tests at all.
I'm having exact same problem and I do not have any E2E tests in my CI, it seems like it is failing on the "test" task looking at the failing agent's output:
```
nx run-many

Run target for multiple listed projects

Options:
      --help               Show help  [boolean]
      --version            Show version number  [boolean]
  -c, --configuration      This is the configuration to use when performing tasks on projects  [string]
  -t, --targets, --target  Tasks to run for affected projects  [array] [required]
      --output-style       Defines how Nx emits outputs tasks logs  [string] [choices: "dynamic", "static", "stream", "stream-without-prefixes"]
      --exclude            Exclude certain projects from being processed  [array] [default: []]
      --parallel           Max number of parallel processes [default is 3]  [string]
      --runner             This is the name of the tasks runner configured in nx.json  [string]
      --verbose            Prints additional information about the commands (e.g., stack traces)  [boolean] [default: false]
      --nx-bail            Stop command execution after the first failed task  [boolean] [default: false]
      --nx-ignore-cycles   Ignore cycles in the task graph  [boolean] [default: false]
      --skip-nx-cache      Rerun the tasks even when the results are available in the cache  [boolean] [default: false]
  -p, --projects           Projects to run. (comma delimited project names and/or patterns)  [array]
      --all                [deprecated] Run the target on all projects in the workspace  [boolean] [default: true]

Examples:
  run-many --target=test                                          Test all projects
  run-many --target=test --projects=proj1,proj2                   Test proj1 and proj2
  run-many --target=test --projects=proj1,proj2 --parallel=2      Test proj1 and proj2 in parallel
  run-many --target=test --projects=*-app --exclude excluded-app  Test all projects ending with `*-app` except `excluded-app`

Find more information and examples at https://nx.dev/nx/run-many

Error: No projects matching: --ci
    at projectsToRun (/home/runner/work/game-companion/game-companion/node_modules/nx/src/command-line/run-many.js:53:23)
    at Object.<anonymous> (/home/runner/work/game-companion/game-companion/node_modules/nx/src/command-line/run-many.js:24:26)
    at Generator.next (<anonymous>)
    at fulfilled (/home/runner/work/game-companion/game-companion/node_modules/tslib/tslib.js:115:62)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
Error: Command execution failed (distributed task execution: ucV6TmkotC). Tasks hashes haven't been recorded.
    at readCompletedTasks (/home/runner/work/game-companion/game-companion/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:7067)
    at invokeTasksUsingRunMany (/home/runner/work/game-companion/game-companion/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:8324)
    at /home/runner/work/game-companion/game-companion/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:6[42](https://github.com/gund/game-companion/actions/runs/3952810529/jobs/6768420810#step:14:43)3
    at Generator.next (<anonymous>)
    at _0x155d[50](https://github.com/gund/game-companion/actions/runs/3952810529/jobs/6768420810#step:14:51) (/home/runner/work/game-companion/game-companion/node_modules/@nrwl/nx-cloud/lib/core/runners/distributed-agent/distributed-agent.impl.js:1:3182)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
``` 

It seems like the main job did not properly send the command and agent received just `nx run-many --ci --code-coverage` but should've been `nx affected --target=test --parallel=3 --ci --code-coverage`.
Not even sure why it's using `nx run-many` if I'm only using `nx affected` in all my CI commands...

My project is open source so you can actually check it out and see [error here](https://github.com/gund/game-companion/actions/runs/3952810529/jobs/6768420810#step:14:73).

Let me know if there is anything I can help you with debugging it or something else.

**Comment by StalkAltan** [comment]
@virtuoushub @gund Please upgrade to latest, you have a separate issue that was caused by some changes made to `run-many`.

@roldengarm is this still an issue for you on latest? Apologies, this issue fell through the cracks.

**Comment by gund** [comment]
> @virtuoushub @gund Please upgrade to latest, you have a separate issue that was caused by some changes made to `run-many`.

Thanks, after update everything seems to run normally in CI!

**Comment by StalkAltan** [comment]
Closing this issue as we have majorly re-worked the way the DTE client works behind the scenes, as of @nrwl/nx-cloud@15.3.0` and later. This version should be compatible with Nx versions above 14.

If you are someone that finds this issue via search, and have a similar problem, please file a **new** issue. Thanks in advance!
